rustls/src/msgs/deframer.rs

```diff
@@ -455,6 +489,24 @@ impl DeframerBuffer for DeframerVecBuffer {
     }
 }
 
+impl DeframerBuffer for DeframerSliceBuffer<'_> {
+    fn advance(&mut self, _num_bytes: usize) {
+        unimplemented!("operation not supported by this type of buffer")
```

Comment:
If we're going to have this, I'd at least like some comments here to describe how we can statically be certain that these won't be invoked.

Reply:
well... we could split `DeframerBuffer` into two traits: One with `filled` and `filled_mut` and another one with `advance` and `unfilled`. Then `DeframerVecBuffer` would implement both but `DeframerSliceBuffer` would only implement the first. Would that work better for you?

Reply:
You're right, that wouldn't be as straightforward.

Another idea I just came up with is to add a const generic parameter to `DeframerBuffer` so it becomes `DeframerBuffer<const QUIC: bool>`.

Then `DeframerSliceBuffer` only implements `DeframerBuffer<false>` and `DeframerVecBuffer` implements `DeframerBuffer<QUIC>` for any value of `QUIC`.

This `QUIC` parameters gets propagated into `MessageDeframer::append_hs` and then we can statically guarantee that `<DeframerSliceBuffer as DeframerBuffer<true>>::copy` is never used as it is not implemented.

Reply:
I think that looks OK -- WDYT @djc?

Eventually I think it would be sensible to refactor the deframer to extract just the parts that QUIC needs to use. Right now a QUIC connection has a full deframer, but only uses the handshake-specific parts of it. That should also make this fall away for TCP-TLS uses.

Reply:
This is a pretty neat solution, yes. Would definitely be interesting to see if we can use a substantially simplified deframer for the QUIC case.